File: Arvores/C/ABP.c
#include <stdio.h> 
#include <stdlib.h>
//Definindo extrutura do Nó
typedef struct No{
    int elemento;
    struct No *esq, *dir;
} No; 
//Cria o nó
No* newNo(int elemento){
    No* no = (No*)malloc(sizeof(No)); //aloca dinamicamente 
    no -> elemento = elemento; // Insere o elemento
    no ->esq = no -> dir= NULL; 
    return no;
}
//Deleta o nó limpando ele da memória
void delNo(No* no){
    free(no);
}
//Criando extrutura dos elementos da árvore 
typedef struct{
    No* raiz;
} ABP;
//Cria a árvore 
ABP* newABP(){
    ABP* abp = (ABP*)malloc(sizeof(ABP));//Aloca dinamicamente
    abp->raiz = NULL;//Inicialmente a raiz é nula 
    return abp;
}

void delABP(ABP* arv){
    //desalocar todos os nós
    free(arv);
}
// Método auxiliar para pesquisar
int pesquisar(ABP* arv, int x){
    return pesquisarRec(x, arv->raiz); // chama o método pricipal
}

int pesquisarRec(int x, No* i){
    int resp;
    if(i==NULL){
        resp= 0;
    }
    else if(x==i->elemento){
        resp = 1;
    }
    else if(x < i->elemento){
        resp = pesquisarRec(x, i->esq);
    }
    else if(x > i->elemento){
        resp = pesquisarRec(x, i->dir);
    }
    return resp;
}

void inserir(ABP* arv, int x){
    inserirRec(x, & arv->raiz);
}

void inserirRec(int x, No** i){
     if(*i == NULL){
        *i = newNo(x);
    }
    else if(x == &((*i))){
        printf("erro ao inserir");
    }
    else if(x < &((*i))){
        (*i)->esq = x;  
    }
    else if(x > &((*i))){
        (*i)->dir =x; 
    }
}

int main()
{
    ABP* arv = newABP();
    inserir(arv, 5);
    inserir(arv, 3);
    printf("%d\n", pesquisar(arv,3));
    delABP(arv);
}

File: Arvores/Java/ABP.java
public class ABP {
    private No raiz;

    public ABP() {
        raiz = null;
    }

    public void caminhar() {
        caminhar(raiz);
        System.out.println();
    }

    public void caminharPre() {
        caminharPre(raiz);
        System.out.println();
    }

    public void caminharPos() {
        caminharPos(raiz);
        System.out.println();
    }


    private void caminhar(No i) {
        if (i != null) {
            caminhar(i.esq);
            System.out.print(i.elemento + " ");
            caminhar(i.dir);
        }
    }

    private void caminharPre(No i) {
        if (i != null) {
            System.out.print(i.elemento + " ");
            caminharPre(i.esq);
            caminharPre(i.dir);
        }
    }

    private void caminharPos(No i) {
        if (i != null) {
        caminharPos(i.esq);
        caminharPos(i.dir);
        System.out.print(i.elemento + " ");
        }
        }
    public void inserir(int x) throws Exception {
        raiz = inserir(x, raiz);
    }

    private No inserir(int x, No i) throws Exception {
        if (i == null) {
            i = new No(x);
        } else if (x < i.elemento) {
            i.esq = inserir(x, i.esq);
        } else if (x > i.elemento) {
            i.dir = inserir(x, i.dir);
        } else {
            throw new Exception("Erro!");
        }
        return i;
    }

    public boolean pesquisar(int x) {
        return pesquisar(x, raiz);
    }

    private boolean pesquisar(int x, No i) {
        boolean resp;
        if (i == null) {
            resp = false;
        } else if (x == i.elemento) {
            resp = true;
        } else if (x < i.elemento) {
            resp = pesquisar(x, i.esq);
        } else {
            resp = pesquisar(x, i.dir);
        }
        return resp;
    }

    public int getAltura(){
        return getAltura(raiz);
    }

    private int getAltura(No i){
        int h = -1; 
        if(i == null){
            int hesq = getAltura(i.esq);
            int hdir = getAltura(i.dir);
            int hmax = hesq > hdir ? hesq : hdir;
            h = hmax + 1; 
        }
        return h; 
    }



}
